sql: shrink SampleReservoir capacity on memory exhaustion

michae2 opened the pull request:
**sql: dynamically shrink sampleAggregator capacity as needed**

To avoid bias, sampleAggregator must always use a capacity <= the
capacity of each samplerProcessor feeding it. This was always implicitly
true before, as every samplerProcessor and sampleAggregator used the
same fixed capacity. But the next few commits will give sampleProcessors
(and sampleAggregators) the ability to dynamically decrease capacity
when out of memory, meaning we now sometimes need to resize
sampleAggregator to keep this invariant true.

This commit does not yet change any behavior because the capacities of
samplerProcessors are still static. Next few commits will change that.

Release note: None

---

**sql: fix memory accounting in SampleReservoir.copyRow**

When overwriting the highest-rank sample with a new one, we might have
*smaller* datums than before, so we should account for memory use
shrinking as well as growing. (Of course, those datums are in a
datumAlloc, so we don't know exactly when they will be GC'd... but this
is the best accounting we can do without hooking into the garbage
collector.)

Release note: None

---

**sql: shrink SampleReservoir capacity on memory exhaustion**

Fixes: #62206

Instead of returning an error when out of memory, make
SampleReservoir.SampleRow dynamically reduce capacity and retry. This
will result in a lower-resolution histogram, but lower-resolution is
better than no histogram at all!

Release note (bug fix): fix histogram generation when table statistics
collection reaches memory limits.

---

**sql: add minimum SampleReservoir capacity**

Below a certain number of samples, the histograms would be so imprecise
that they are not worth generating. Add the ability to set this
threshold in SampleReservoir. This commit sets this threshold to the
max number of histogram buckets, which was picked because it seemed
to make sense and not for any mathematical or empirical reason.

Release note: None

michae2 committed:
sql, stats: shrink SampleReservoir capacity on memory exhaustion

Fixes: #62206

Instead of returning an error when out of memory, make
SampleReservoir.SampleRow dynamically reduce capacity and retry. Fewer
samples will result in a less accurate histogram, but less accurate is
probably still better than no histogram at all, up to a point.

If the number of samples falls below a minimum threshold, then give up
and disable histogram collection, as we were doing originally, rather
than using a wildly inaccurate histogram.

Also fix some memory accounting in SampleReservoir.copyRow.

Release note (performance improvement): continue to generate histograms
when table statistics collection reaches memory limits, instead of
disabling histogram generation.

michae2 committed:
sql, stats: shrink sampleAggregator capacity to match child capacity

To avoid bias, sampleAggregator must always use a capacity <= the
capacity of each child samplerProcessor feeding it. This was always
implicitly true before, as every samplerProcessor and sampleAggregator
used the same fixed capacity. But the next few commits will give
sampleProcessors (and sampleAggregators) the ability to dynamically
shrink capacity when out of memory, meaning we now sometimes need to
resize sampleAggregator to keep this invariant true.

(Resizing sampleAggregator really means resizing the underlying
SampleReservoir, so most of the changes are there.)

This commit does not yet change any behavior, because the capacities of
samplerProcessors are still static. Next few commits will change that.

Also factor the giant closure out of TestSampleAggregator into an
external function, to make table-driven testing easier.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65491)
<!-- Reviewable:end -->

michae2 commented:
I'm still working on testing, but this is probably RFAL.

yuzefovich reviewed (commented):
I'll leave it to @rytaft to review it much closer, just a couple of quick comments from me.

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-Ma69Kpm6RPaMsg5gJNT:biwqxj4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/rowexec/sampler.go, line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma6954L20dRgIqbw0rq:-Ma6954L20dRgIqbw0rr:b8x6og7) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/rowexec/sampler.go#L115)):*
> ```Go
> 
> 	sampleSize, minSampleSize := int(spec.SampleSize), int(spec.MinSampleSize)
> 	if minSampleSize < 1 || minSampleSize > sampleSize {
> ```

super nit: this logic is duplicated in two places. Might be worth extracting into a helper.
___
*[pkg/sql/stats/row_sampling.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma67oZmDF--QhIC9CMM:-Ma67oZmDF--QhIC9CMN:b-2200vh) ([raw file](https://github.com/cockroachdb/cockroach/blob/863a79c89592702fc9af6ccfb28017db573739b0/pkg/sql/stats/row_sampling.go#L134)):*
> ```Go
> 		// for the additional memory used after copying inside copyRow.
> 		if sr.memAcc != nil {
> 			if err := sr.memAcc.Grow(ctx, int64(len(rowCopy))*int64(rowCopy[0].Size())); err != nil {
> ```

This line seems suspicious. If we have a row of `INT, BYTES, DECIMAL`, we will account for `3 x INT`. Then in `copyRow` our `beforeSize` will be `rowCopy[i].Size` i.e. each of the datums separately.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/stats/row_sampling.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma67oZmDF--QhIC9CMM:-Ma9fZvA4FVBbzaNYyxW:bxz58pi) ([raw file](https://github.com/cockroachdb/cockroach/blob/863a79c89592702fc9af6ccfb28017db573739b0/pkg/sql/stats/row_sampling.go#L134)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This line seems suspicious. If we have a row of `INT, BYTES, DECIMAL`, we will account for `3 x INT`. Then in `copyRow` our `beforeSize` will be `rowCopy[i].Size` i.e. each of the datums separately.
</blockquote></details>

I believe at this point `rowCopy[0].Size()` will only be `EncDatumOverhead` because we have not yet set the values to anything, so this should be equivalent to the sum of `Size()` of each element of `rowCopy`. But this line confused me, too. I will change it to use `rowCopy.Size()`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/stats/row_sampling.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma67oZmDF--QhIC9CMM:-Ma9ldPK78JFl3EmuptE:b-ep2znb) ([raw file](https://github.com/cockroachdb/cockroach/blob/863a79c89592702fc9af6ccfb28017db573739b0/pkg/sql/stats/row_sampling.go#L134)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I believe at this point `rowCopy[0].Size()` will only be `EncDatumOverhead` because we have not yet set the values to anything, so this should be equivalent to the sum of `Size()` of each element of `rowCopy`. But this line confused me, too. I will change it to use `rowCopy.Size()`.
</blockquote></details>

Oh yeah, indeed, we haven't performed a copy yet, nvm then.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Very cool! Needs some tests, but definitely seems like it's on the right track.

> Reviewed 6 of 6 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-Ma9rya83Wl3jjlTJmze:bnga2it)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/rowexec/sample_aggregator.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9Eh6mEwZejt-LZaXv:-Ma9Eh6mEwZejt-LZaXw:bb7xg9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sample_aggregator.go#L287)):*
> ```Go
> 				// Inverted sample row.
> 				// Retain the rows with the top ranks.
> 				if capacity, err := row[s.numRowsCol].GetInt(); err == nil {
> ```

This could use an update to the comment explaining what's happening.
___
*[pkg/sql/rowexec/sample_aggregator.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9FDtq6cmauDuKMVR1:-Ma9FDtq6cmauDuKMVR2:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sample_aggregator.go#L309)):*
> ```Go
> 			// Sample row.
> 			// Retain the rows with the top ranks.
> 			if capacity, err := row[s.numRowsCol].GetInt(); err == nil {
> ```

ditto
___
*[pkg/sql/rowexec/sampler_test.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9HKuBC94KG14QjDzq:-Ma9HKuBC94KG14QjDzr:bc0tdlw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sampler_test.go#L102)):*
> ```Go
> 		}
> 		for i := 2; i < len(outTypes); i++ {
> 			if i != 3 /* num rows col */ && !row[i].IsNull() {
> ```

nit: I'd just make a const var called `numRowsCol`. While you're at it, it would be awesome if you could improve this test by making const vars for the other columns too.
___
*[pkg/sql/stats/row_sampling.go, line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9IIEw9T-hJE9VyJ3J:-Ma9IIEw9T-hJE9VyJ3K:b-aarud9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/stats/row_sampling.go#L82)):*
> ```Go
> }
> 
> func (sr *SampleReservoir) Cap() int {
> ```

nit: needs a comment
___
*[pkg/sql/stats/row_sampling.go, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9Iq2JEG_wJ-0rQS1Q:-Ma9Iq2L-wHiS90TFsVM:b-lopmct) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/stats/row_sampling.go#L117)):*
> ```Go
> 		samp := heap.Pop(sr).(SampledRow)
> 		if sr.memAcc != nil {
> 			sr.memAcc.Shrink(ctx, int64(samp.Row.Size()))
> ```

I'm not sure we can actually call shrink here without allocating a completely new array. Even though `samples` no longer references these rows, I don't think they can be garbage collected until all references to the underlying array are gone, at which point the entire array will be GC'd.
___
*[pkg/sql/stats/row_sampling.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LScKAEGUdMRGhF4R:-Ma9LScL4G25jKsguvUy:bd0vnw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L124)):*
> ```Go
> }
> 
> func (sr *SampleReservoir) trySampleRow(
> ```

Would be good to add a function comment here 
___
*[pkg/sql/stats/row_sampling.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LWcbDhOBHfDQY305:-Ma9LWcbDhOBHfDQY306:b-nqzkko) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L151)):*
> ```Go
> 	if len(sr.samples) > 0 && rank < sr.samples[0].Rank {
> 		if err := sr.copyRow(ctx, evalCtx, sr.samples[0].Row, row); err != nil {
> 			// XXX At this point sr.samples[0].Row could have a mix of old and new
> ```

What's the XXX for?
___
*[pkg/sql/stats/row_sampling.go, line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LlugBqlJVgBAfCJY:-Ma9LlugBqlJVgBAfCJZ:bzfoomw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L176)):*
> ```Go
> 		}
> 		// We've used too much memory. Remove the highest-rank sample and try again.
> 		sr.Resize(ctx, len(sr.samples)-1)
> ```

I think we probably want to be more aggressive here -- otherwise we'll keep hitting the memory limit. Instead of just removing one sample at a time, maybe we should cut the number of samples in half (or reduce by some other fraction of the total). (This will also amortize the cost of creating a new array on each Resize, which as I mention above, is something I think you need to do to make this work correctly).
___
*[pkg/sql/stats/row_sampling.go, line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9rMn46JOEcAwmVZRQ:-Ma9rMn5-bynVNPSksC2:b2tg466) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/stats/row_sampling.go#L80)):*
> ```Go
> 	}
> 	if minNumSamples < 1 {
> 		return errors.AssertionFailedf("minNumSamples (%d) < 1", numSamples)
> ```

numSamples -> minNumSamples
___
*[pkg/sql/stats/row_sampling.go, line 191 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9rjfKDj_g1l-5Ka_y:-Ma9rjfKDj_g1l-5Ka_z:b5j2v5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/stats/row_sampling.go#L191)):*
> ```Go
> 		err := sr.trySampleRow(ctx, evalCtx, row, rank)
> 		if err == nil || !sqlerrors.IsOutOfMemoryError(err) ||
> 			len(sr.samples) <= 1 || cap(sr.samples) <= sr.minNumSamples {
> ```

Do you still need to check if `len(sr.samples) <= 1`? Isn't that handled by minNumSamples?


<!-- Sent from Reviewable.io -->

rytaft commented:
I think you also need to run `make bazel-generate`





<!-- Sent from Reviewable.io -->

michae2 committed:
sql, stats: shrink sampleAggregator capacity before histogram generation

SampleAggregator can also run out of memory when gathering the datums
for histogram generation. Push this datum gathering down into
SampleReservoir so that we can use the same dynamic capacity-shrinking
strategy as used in SampleRow to reclaim some memory and try again.

Release note: None

michae2 reviewed (commented):
Well, still no test but I've:
* addressed the comments (thank you)
* added some logging
* refactored the retry in `SampleReservoir`
* fixed a spot I missed in `sampleAggregator.generateHistogram`

Test is coming, I swear, but pushing this to see what TeamCity thinks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-MabKldj4mL6AONDR-nD:barc3dx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9Eh6mEwZejt-LZaXv:-MabCq0NCvxYkQEsoCER:brg8r7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sample_aggregator.go#L287)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This could use an update to the comment explaining what's happening.
</blockquote></details>

OK, how's this?
___
*[pkg/sql/rowexec/sample_aggregator.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9FDtq6cmauDuKMVR1:-MabCwYB7dVq_hYrvXvG:b2nhktp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sample_aggregator.go#L309)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

OK.
___
*[pkg/sql/rowexec/sampler.go, line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma6954L20dRgIqbw0rq:-MabHI5aDciM9LcVNTDY:b9eq5vn) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/rowexec/sampler.go#L115)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

super nit: this logic is duplicated in two places. Might be worth extracting into a helper.
</blockquote></details>

OK, I moved it into `SampleReservoir.Init`.
___
*[pkg/sql/rowexec/sampler_test.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9HKuBC94KG14QjDzq:-MabDEfy4-aEM6opKL2O:bajlnx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sampler_test.go#L102)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: I'd just make a const var called `numRowsCol`. While you're at it, it would be awesome if you could improve this test by making const vars for the other columns too.
</blockquote></details>

OK, this should be a little better now.
___
*[pkg/sql/stats/row_sampling.go, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9Iq2JEG_wJ-0rQS1Q:-MabDOHf7ePGRnQrKSzK:b7lp2m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/stats/row_sampling.go#L117)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'm not sure we can actually call shrink here without allocating a completely new array. Even though `samples` no longer references these rows, I don't think they can be garbage collected until all references to the underlying array are gone, at which point the entire array will be GC'd.
</blockquote></details>

Oh crap, good catch! Done.
___
*[pkg/sql/stats/row_sampling.go, line 134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma67oZmDF--QhIC9CMM:-MabE0PU0i82_6JkVpEn:b-vfx0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/863a79c89592702fc9af6ccfb28017db573739b0/pkg/sql/stats/row_sampling.go#L134)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Oh yeah, indeed, we haven't performed a copy yet, nvm then.
</blockquote></details>

OK, this is now using `rowCopy.Size()`.
___
*[pkg/sql/stats/row_sampling.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LScKAEGUdMRGhF4R:-MabGge99d4x4mhgnLRV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L124)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Would be good to add a function comment here 
</blockquote></details>

Done.
___
*[pkg/sql/stats/row_sampling.go, line 151 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LWcbDhOBHfDQY305:-MabGjIn83pRWO1ES0gL:b-o6372o) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L151)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What's the XXX for?
</blockquote></details>

This was a type of comment we used at a previous company. Changed it to `WARNING`.
___
*[pkg/sql/stats/row_sampling.go, line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9LlugBqlJVgBAfCJY:-MabGxQx8I0mA02D6qjr:bsywv9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d12aece747e52eda3cf29f1573eae7012726526/pkg/sql/stats/row_sampling.go#L176)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think we probably want to be more aggressive here -- otherwise we'll keep hitting the memory limit. Instead of just removing one sample at a time, maybe we should cut the number of samples in half (or reduce by some other fraction of the total). (This will also amortize the cost of creating a new array on each Resize, which as I mention above, is something I think you need to do to make this work correctly).
</blockquote></details>

Yes, you're right. Done.
___
*[pkg/sql/stats/row_sampling.go, line 80 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9rMn46JOEcAwmVZRQ:-MabHSCv06iYQ8Fu6hR9:brtxory) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/stats/row_sampling.go#L80)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

numSamples -> minNumSamples
</blockquote></details>

Ah! Good catch! But I ended up changing it anyway. :smile:
___
*[pkg/sql/stats/row_sampling.go, line 191 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9rjfKDj_g1l-5Ka_y:-MabHypS5QgRjjM2Pezn:b-zfq7z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/13aa166f869c62a0b969d599cc244c454351ae5b/pkg/sql/stats/row_sampling.go#L191)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you still need to check if `len(sr.samples) <= 1`? Isn't that handled by minNumSamples?
</blockquote></details>

It is in almost every case, but in the (extremely unlikely) case where we hit a memory limit with only 0 or 1 row sampled the `len` check would matter. (The memory limit would have to be very very low for this to happen.)


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Latest force push:

* New commit adding memory limit testing to `row_sampling_test.go`.
* This uncovered bugs in `SampleReservoir.Pop` and `SampleReservoir.MaybeResize` which were fixed.
* Also fixed some small compilation errors and a few comments.

I believe there will be one more commit and force push to add some testing around `samplerProcessor` and `sampleAggregator`, so you don't have to look yet (though you're welcome to).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-MagAMK3CiMEb0VkKiaB:b-wu4cxu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 4 of 8 files at r5, 2 of 2 files at r6, 5 of 8 files at r9, 1 of 2 files at r10, 3 of 3 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator.go, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Ma9Eh6mEwZejt-LZaXv:-MakUu-i4Wcdm-7bBuJx:bsrfkwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2008f60773317cb9ff4d46c145d69dc144bffd1d/pkg/sql/rowexec/sample_aggregator.go#L287)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

OK, how's this?
</blockquote></details>

Looks good
___
*[pkg/sql/rowexec/sample_aggregator.go, line 379 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakVFzL0_7Erim4oGsT:-MakVFzL0_7Erim4oGsU:b-4t1qff) ([raw file](https://github.com/cockroachdb/cockroach/blob/dda2519d186c236b855deaf18b7a49bb8f2e4350/pkg/sql/rowexec/sample_aggregator.go#L379)):*
> ```Go
> 		if sr.MaybeResize(ctx, int(capacity)) {
> 			log.Infof(
> 				ctx, "histogram samples reduced from %d to %d to match sample processor",
> ```

nit: sample -> sampler
___
*[pkg/sql/rowexec/sample_aggregator.go, line 401 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakXKvUB1uoCcFSsxyY:-MakWuuwBVW0o_jnL8HK:b-btkc9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/rowexec/sample_aggregator.go#L401)):*
> ```Go
> 	} else if sr.Cap() != prevCapacity {
> 		log.Infof(
> 			ctx, "histogram samples reduced from %d to %d due to excessive memory utilization",
> ```

are we logging this same message multiple times? or do these all cover different cases?
___
*[pkg/sql/stats/row_sampling.go, line 60 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakXcgg45UCM796FTNL:-MakX_G3-SoXYASQJI4l:bxtnge2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling.go#L60)):*
> ```Go
> 	// minNumSamples is the minimum K needed for sampling to be meaningful. If the
> 	// reservoir capacity would fall below this, SampleRow will err instead of
> 	// decreasing K further.
> ```

nit: Here and below throughout this file you reference "K", but there's no variable named K, and I think it just adds confusion. I'd just say "capacity" (if that's the intended meaning) instead of "K".
___
*[pkg/sql/stats/row_sampling_test.go, line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakaMDz3zmuJ5ysRSNc:-MakaMDz3zmuJ5ysRSNd:baf6vli) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling_test.go#L45)):*
> ```Go
> 		} else if sr.Cap() != prevCapacity {
> 			t.Logf(
> 				"samples reduced from %d to %d during SampleRow",
> ```

Instead of just logging, maybe assert that the number of samples was reduced by an expected amount?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Latest push:
* rebases on master tip
* adds some testing for `samplerProcessor` and `sampleAggregator`
* fixes a bug in `SampleReservoir.retryMaybeResize` that this testing caught
* no longer silently drops errors returned by `SampleReservoir.Init`
* tweaks some comments and commit messages slightly

I think this is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-Mb_9ku6CilsRmrk5p4C:b-olkk8e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator.go, line 401 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakXKvUB1uoCcFSsxyY:-Mb_CWX81LQMjWLWxKss:b-6wnpps) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/rowexec/sample_aggregator.go#L401)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

are we logging this same message multiple times? or do these all cover different cases?
</blockquote></details>

These cover two different cases:
1. A child `samplerProcessor` ran out of memory, and this `sampleAggregator` has to shrink to match the capacity of that child.
2. This `sampleAggregator` itself ran out of memory.
___
*[pkg/sql/stats/row_sampling.go, line 60 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakXcgg45UCM796FTNL:-Mb_BJQb7V4FYXLuhwl2:b5nose0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling.go#L60)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: Here and below throughout this file you reference "K", but there's no variable named K, and I think it just adds confusion. I'd just say "capacity" (if that's the intended meaning) instead of "K".
</blockquote></details>

"K" comes from the big comment above the definition of `SampleReservoir`. I've changed some of the comments so that they say "reservoir capacity (K)" instead of just "K", is that better?
___
*[pkg/sql/stats/row_sampling_test.go, line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakaMDz3zmuJ5ysRSNc:-Mb_C2WqCEchu2uJu1W4:bnkz21l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling_test.go#L45)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Instead of just logging, maybe assert that the number of samples was reduced by an expected amount?
</blockquote></details>

It's a good point, but I'm worried that specifying an exact amount would cause the test to flake if memory usage changed slightly and the `SampleReservoir` shrunk more or less than it did before.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Sigh, sorry for one more push. As it turns out, `SampleReservoir` needs to support a capacity of 0. (`samplerProcessor` and `sampleAggregator` use this when histograms are disabled). So just a slight change to `SampleReservoir.Init` and `SampleReservoir.retryMaybeResize` is needed. This should make the tests happy.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-MbbfmcjANoplc3QddzW:byvdstp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
This is getting really close!

It seems like there are a lot of commits -- can you squash any of them? Also make sure that each commit builds/passes tests individually.

Correct me if I'm wrong, but I don't see any tests that specifically verify the goal of this PR to degrade gracefully with tightening memory conditions. So instead of just verifying that we do or don't run out of memory in a certain test case, it would be good to check that in cases where we would have run out of memory previously, now we succeed, but the number of samples is reduced to an expected number (both in the sampler and the sampleAggregator). You can use `make stress` to make sure that you're not introducing flakes.

> Reviewed 9 of 9 files at r12, 1 of 1 files at r13, 2 of 4 files at r14, 5 of 8 files at r15, 1 of 2 files at r16, 3 of 3 files at r17, 1 of 1 files at r18, 1 of 1 files at r20, 4 of 8 files at r24, 1 of 4 files at r25, 1 of 2 files at r26, 3 of 3 files at r27, 1 of 1 files at r28, 1 of 1 files at r30.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-MbbbOJl1NPlqsMSEWKb:b-fz6poe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 305 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbPrIB8BpuUoapkyum:-MbbPrIB8BpuUoapkyun:bo33492) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sample_aggregator_test.go#L305)):*
> ```Go
> 	)
> 
> 	inputRows = [][]int{
> ```

would help to add a comment to explain what the following cases are testing
___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 348 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbOyDFEhnJ9abvuRSI:-MbbOyDG5ZIxIcasw7XY:b-wd3wdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sample_aggregator_test.go#L348)):*
> ```Go
> 	runSampleAggregator(
> 		t, server, sqlDB, kvDB, st, &evalCtx, 1<<15 /* memLimitBytes */, false, /* expectOutOfMemory */
> 		2 /* numSamples */, 4 /* aggNumSamples */, inputRows, expected,
> ```

are these supposed to be different?
___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbjG5U5Et15V-kpu1d:-MbbjG5U5Et15V-kpu1e:b-2vhdsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/rowexec/sample_aggregator_test.go#L152)):*
> ```Go
> 	spec := &execinfrapb.SampleAggregatorSpec{
> 		SampleSize:       aggNumSamples,
> 		MinSampleSize:    2,
> ```

you could pass this as a param too
___
*[pkg/sql/rowexec/sampler_test.go, line 208 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbR3Nb4FruuGz4ih2j:-MbbR3Nc-eNeMB2RQzat:b-vpssav) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sampler_test.go#L208)):*
> ```Go
> 	runSampler(t, numRows, numSamples, numSamples/2, memLimitBytes, true /* expectOutOfMemory */)
> 	runSampler(t, numRows, numSamples, numSamples/3, memLimitBytes, true /* expectOutOfMemory */)
> 	runSampler(t, numRows, numSamples, numSamples/4, memLimitBytes, false /* expectOutOfMemory */)
> ```

what are these tests showing?
___
*[pkg/sql/stats/row_sampling.go, line 60 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakXcgg45UCM796FTNL:-MbbN6R40vWrYhmgyFNb:buzllzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling.go#L60)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

"K" comes from the big comment above the definition of `SampleReservoir`. I've changed some of the comments so that they say "reservoir capacity (K)" instead of just "K", is that better?
</blockquote></details>

ah got it - thanks!
___
*[pkg/sql/stats/row_sampling.go, line 131 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Mbbgt_LDIsckSAqeGiz:-Mbbgt_LDIsckSAqeGj-:b-rkuaad) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/stats/row_sampling.go#L131)):*
> ```Go
> // introducing bias. Increasing capacity would cause later rows to be
> // over-represented relative to earlier rows.)
> func (sr *SampleReservoir) MaybeResize(ctx context.Context, k int) bool {
> ```

mention in the comment what the return param means (or just remove the return param -- is it even used?)
___
*[pkg/sql/stats/row_sampling.go, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbhN7y3f4BWcMjRn_z:-MbbhN7z4U5h1u-NeQJS:bebonpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/stats/row_sampling.go#L152)):*
> ```Go
> // retryMaybeResize tries to execute a memory-allocating operation, shrinking
> // the capacity of the reservoir (K) as necessary until the operation succeeds
> // or fails with a non-out-of-memory error.
> ```

nit: this implies it would never fail with an out of memory error, but that doesn't seem to be the case
___
*[pkg/sql/stats/row_sampling_test.go, line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakaMDz3zmuJ5ysRSNc:-MbbRK5H5lCY5tFB6sWU:btyplco) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling_test.go#L45)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It's a good point, but I'm worried that specifying an exact amount would cause the test to flake if memory usage changed slightly and the `SampleReservoir` shrunk more or less than it did before.
</blockquote></details>

Maybe you could assert that it's within a range? I think it would be better to deal with flakes and tweak the test as needed than to have it silently fail. You can check for possible flakes by running `make stress` on the test. If it runs a large number of times without failing it should be good to go.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 348 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbOyDFEhnJ9abvuRSI:-Mbc3IQL9OOr3GAge5BU:bnyxaks) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sample_aggregator_test.go#L348)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

are these supposed to be different?
</blockquote></details>

Here we're varying the memory limit, and below we're varying numSamples and aggNumSamples. But it's not really clear. Let me clean this up and add some comments.
___
*[pkg/sql/rowexec/sampler_test.go, line 208 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbR3Nb4FruuGz4ih2j:-MbboGtZ5gpQVkWg3ePX:bd7xw1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sampler_test.go#L208)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what are these tests showing?
</blockquote></details>

This is the closest to the testing you describe (though not exactly the same, and not as clear): it holds the memory limit fixed while decreasing `minNumSamples`, the required number of samples, until finding the point at which we can stay within the memory limit. `minNumSamples = numSamples` is equivalent to the old behavior (fail if we do not have enough memory for the requested capacity). From there we decrease the require number of samples until we're inside the memory limit.

But I think what you described (holding number of samples fixed while varying the memory limit) is a better test, so I will add that.
___
*[pkg/sql/stats/row_sampling_test.go, line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakaMDz3zmuJ5ysRSNc:-Mbc32Ni3Oip_QAu8u8A:b-mdn7sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling_test.go#L45)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe you could assert that it's within a range? I think it would be better to deal with flakes and tweak the test as needed than to have it silently fail. You can check for possible flakes by running `make stress` on the test. If it runs a large number of times without failing it should be good to go.
</blockquote></details>

Good points, you are right. I will do that.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/sampler_test.go, line 208 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbR3Nb4FruuGz4ih2j:-MbcQU2f91a5SDhlX8qe:bzc2yfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sampler_test.go#L208)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This is the closest to the testing you describe (though not exactly the same, and not as clear): it holds the memory limit fixed while decreasing `minNumSamples`, the required number of samples, until finding the point at which we can stay within the memory limit. `minNumSamples = numSamples` is equivalent to the old behavior (fail if we do not have enough memory for the requested capacity). From there we decrease the require number of samples until we're inside the memory limit.

But I think what you described (holding number of samples fixed while varying the memory limit) is a better test, so I will add that.

</blockquote></details>

This seems like a good test too, maybe just add a comment to explain


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Ok, latest push squashes this all into three commits, and enhances the testing a bit, hopefully to good effect. :slightly_smiling_face: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-Mbe9DpYAe7kJBVO39lM:bqn5yic)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 305 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbPrIB8BpuUoapkyum:-Mbe8kx02QaX3WcDMXBh:b1z80l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sample_aggregator_test.go#L305)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

would help to add a comment to explain what the following cases are testing
</blockquote></details>

Done (down in the table of test cases).
___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbjG5U5Et15V-kpu1d:-Mbe8hJDCojKVEtl_Twr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/rowexec/sample_aggregator_test.go#L152)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

you could pass this as a param too
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/sampler_test.go, line 208 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbR3Nb4FruuGz4ih2j:-Mbe8sUa-ZecmGPnRof4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sampler_test.go#L208)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This seems like a good test too, maybe just add a comment to explain
</blockquote></details>

Done.
___
*[pkg/sql/stats/row_sampling.go, line 131 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-Mbbgt_LDIsckSAqeGiz:-Mbe8vAf3F828IGF0h_6:bdxs3eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/stats/row_sampling.go#L131)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

mention in the comment what the return param means (or just remove the return param -- is it even used?)
</blockquote></details>

Done. (It is only used to simplify some logging in `sample_aggregator.go`.)
___
*[pkg/sql/stats/row_sampling.go, line 152 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbhN7y3f4BWcMjRn_z:-Mbe969R5NwtK598_kES:b-9o7g4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ce7580c46fcbf5c2744d9c33ca0a811c685a0f7/pkg/sql/stats/row_sampling.go#L152)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: this implies it would never fail with an out of memory error, but that doesn't seem to be the case
</blockquote></details>

You're right, I changed the comment a bit.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)

___
*[pkg/sql/rowexec/sample_aggregator_test.go, line 348 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MbbOyDFEhnJ9abvuRSI:-MbeA59a0b1w4Fpav-_p:bhbb1b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfaf03b01bbab68e5ee7b9bd3efef627fc0a67bc/pkg/sql/rowexec/sample_aggregator_test.go#L348)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Here we're varying the memory limit, and below we're varying numSamples and aggNumSamples. But it's not really clear. Let me clean this up and add some comments.
</blockquote></details>

(Done.)
___
*[pkg/sql/stats/row_sampling_test.go, line 45 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-MakaMDz3zmuJ5ysRSNc:-MbeAC9x5rxaw9TW7Od5:b6yfusk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc3a809201ef2e33e493bb15671d3d729e31b2f6/pkg/sql/stats/row_sampling_test.go#L45)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Good points, you are right. I will do that.
</blockquote></details>

I ended up checking it exactly, which does seem fine.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
nit: I wouldn't call this a bug fix (in your release note). "performance improvement" seems more accurate.

Otherwise, <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Nice work getting this over the finish line!

Let's let this bake on master for a few weeks before backporting to 21.1 and maybe 20.2.

> Reviewed 11 of 11 files at r31, 10 of 10 files at r32, 4 of 4 files at r33.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65491#-:-MbfZSU-CQrkx2-6W3xg:bq7nkac)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

michae2 commented:
TFTR!

bors r+
